### PR TITLE
I156

### PR DESCRIPTION
**Pull request by hillalex**
Please review look and feel as well as code changes!

* add extra guidance to program data help text to explain expectations
* validate program data rows to make sure for each year we have either sex aggregated or sex dis-aggregated data but never both
* don't allow user to proceed if program data is in an invalid state, and display error alert on the review input data tab
* drop the age group column for program data (but hard-code it as 15-99 before passing to model fitting)

Fixes MRC-156 and MRC-162

### CODE REVIEW

**Comment by hillalex** [comment]
Can also merge: https://github.com/mrc-ide/shiny90_sample_files/pull/2
and return sample_files submodule to point to master once ^ is merged

File: src/server/survey_and_program_data.R [hunk 1]
```diff
@@ -61,6 +61,27 @@ programDataHeaders <- list(tot= "Total Tests",
...
 )
 
+validateProgramData <- function(df) {
+
+    if (is.null(df))
+        return(FALSE)
+
+    validateYear <- function(givenYear) {
+
+        rows <- df[!is.na(df$year) & df$year == givenYear,]
```
Comment (r-ash) [inline src/server/survey_and_program_data.R]: We could use which here (may be slightly more readable?)

```
rows <- df[which(df$year == givenYear), ]
```

**Comment by r-ash** [comment]
> * don't allow user to proceed if program data is in an invalid state, and display error alert on the review input data tab

This works with a new file. If I load from a zip and then edit the program data in place to break it (i.e. have 2 2010 rows with "both" sex) then I do not get a warning on the "Review input data" page. Is that a thing we can fix?

I wonder if it worth adding a test case too for multiple years.

```
testthat::test_that("providing both sex aggregated and disaggregated data for multiple years is valid", {
  
  df = data.frame(c(2010, 2011, 2011), c("both", "male", "female"))
  colnames(df) <- c("year", "sex")
  
  result <- validateProgramData(df)
  
  testthat::expect_true(result)
})
```

File: src/server/survey_and_program_data.R [hunk 2]
```diff
@@ -61,6 +61,27 @@ programDataHeaders <- list(tot= "Total Tests",
...
+
+        if (nrow(rows) == 0) return(TRUE)
```
Comment (r-ash) [inline src/server/survey_and_program_data.R]: I found the series of ifs a little hard to follow on first looking, though I don't know if there is a neat way to make it look cleaner in R. There is a case_when function in dplyr which may make it a little easier to read. What do you think?

```
dplyr::case_when(
          nrow(rows) == 0 ~ TRUE,
          nrow(rows) == 1 ~ identical(as.character(rows$sex),c("both")),
          nrow(rows) == 2 ~ identical(sort(as.character(rows$sex)), sort(c("male", "female"))),
          TRUE ~ FALSE )
```

The `TRUE ~ FALSE` bit is not very neat.

**Comment by hillalex** [comment]
See if you prefer this simplified logic. Good shout on the multi-year test, have added it as suggested. I can't replicate the bug you mention - when loading a zip and editing program data I get the invalid data error as expected... could you post exact steps to replicate?

File: src/server/survey_and_program_data.R [hunk 3]
```diff
@@ -68,14 +68,11 @@ validateProgramData <- function(df) {
...
+        rows <- df[which(df$year == givenYear),]
 
-        if (nrow(rows) == 0) return(TRUE)
-        if (nrow(rows) > 2) return(FALSE)
-        if (nrow(rows) == 1 & identical(as.character(rows$sex),c("both"))) return(TRUE)
-        if (nrow(rows) == 2 & identical(sort(as.character(rows$sex)), sort(c("male", "female")))) return(TRUE)
-
-        FALSE
+        nrow(rows) == 0 |
+        identical(as.character(rows$sex),c("both")) |
```
Comment (hillalex) [inline src/server/survey_and_program_data.R]: realised this implicitly checks the number of rows, so no need to check `nrow` explicitly

Comment (r-ash) [reply src/server/survey_and_program_data.R]: Very nice!

**Comment by r-ash** [comment]
Issue mentioned not seeing the alert when creating invalid rows was because I was checking with dupe years 2005 and we're only checking 2010-2018. We'll update to get the years from the data frame instead of it being hard coded.

File: src/server/survey_and_program_data.R [hunk 4]
```diff
@@ -162,13 +194,14 @@ surveyAndProgramData <- function(input, output, state, spectrumFilesState) {
 
     output$hot_program <- rhandsontable::renderRHandsontable({
         rhandsontable::rhandsontable(state$program_data_human_readable(), rowHeaders = NULL, stretchH = "all") %>%
-            rhandsontable::hot_col("Country or region", readOnly = TRUE) %>%
+            rhandsontable::hot_col("Country or region") %>%
```
Comment (r-ash) [inline src/server/survey_and_program_data.R:179]: This is kind of scope creep of this issue but I wonder if we can make the same fix with the survey data. We're seeing the same problem where when editing in place if I add a new row the country is blank but I cannot edit the field.

Alternatively I'm happy to put in another ticket to fix this separately.

Comment (hillalex) [reply src/server/survey_and_program_data.R:179]: I think lets open a new ticket for this, because at the moment we have no validation of survey data beyond validating that some exists. So lets open a new ticket for adding minimal validation of survey data (just checking country name) and making the editing in place actually work (i.e. setting relevant factors for age and sex, plus making country editable)!

https://vimc.myjetbrains.com/youtrack/issue/mrc-167

**Comment by hillalex** [comment]
Just realised that the travis build steps are not being quite clever enough! They're cloning the sample_files repo down as master each time, so not tracking which commit the submodule is really pointing to. Hence why we have 3 failing tests here (to pass they need the right version of the submodule.) I'll try fix up the travis config so we can see a passing build.

File: src/server/survey_and_program_data.R (hunk 4 above)
Comment (r-ash) [reply src/server/survey_and_program_data.R:179]: Great okay, thanks for that.

**Comment by r-ash** [comment]
To address https://vimc.myjetbrains.com/youtrack/agiles/103-9/104-78?issue=mrc-162 we want to remove the age group column from the program data. 

As we no longer use built in program data calling `select_prgrmdata` is unnecessary as we always get empty program data. We should add a function `createEmptyProgramData` like `createEmptySurveyData` that returns

```
    prg_dat <- data.frame(country = cnt, 
                          year = 2010:2018, tot = NA, totpos = NA,
                          vct = NA, vctpos = NA, anc = NA, ancpos = NA,
                          sex = 'both')
```

Then ensure that age group is not shown on the program data page.
